File: backend/app/services/section_preferences.py
"""Generic configurable-section preferences (#821).

Shared merge/normalize/coerce logic behind both the Home screen sections
(#584) and the Insights page sections (#821). Each caller supplies its own
valid-key whitelist, default order, and optional locked keys (sections that
must always stay enabled, e.g. the Insights main feed).
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from typing import TypedDict


class SectionPreference(TypedDict):
    key: str
    enabled: bool


def _coerce_section(
    raw: object,
    valid_keys: frozenset[str],
    locked_keys: frozenset[str],
) -> SectionPreference | None:
    if not isinstance(raw, dict):
        return None
    key = raw.get("key")
    enabled = raw.get("enabled")
    if not isinstance(key, str):
        return None
    key = key.strip()
    if key not in valid_keys:
        return None
    if not isinstance(enabled, bool):
        return None
    if key in locked_keys:
        enabled = True
    return {"key": key, "enabled": enabled}


def merge_sections(
    stored: Sequence[object] | None,
    *,
    defaults: Sequence[SectionPreference],
    valid_keys: frozenset[str],
    locked_keys: frozenset[str] = frozenset(),
) -> list[SectionPreference]:
    """Resolve stored preferences with defaults for missing or unknown keys."""
    if not stored:
        return [section.copy() for section in defaults]

    merged: list[SectionPreference] = []
    seen: set[str] = set()

    for raw in stored:
        section = _coerce_section(raw, valid_keys, locked_keys)
        if section is None or section["key"] in seen:
            continue
        merged.append(section)
        seen.add(section["key"])

    for default in defaults:
        if default["key"] not in seen:
            merged.append(default.copy())
            seen.add(default["key"])

    return merged


def normalize_sections(
    sections: Sequence[object] | None,
    *,
    valid_keys: frozenset[str],
    locked_keys: frozenset[str] = frozenset(),
) -> list[SectionPreference] | None:
    """Validate and dedupe a PATCH payload; empty list is allowed."""
    if sections is None:
        return None

    normalized: list[SectionPreference] = []
    seen: set[str] = set()

    for raw in sections:
        section = _coerce_section(raw, valid_keys, locked_keys)
        if section is None or section["key"] in seen:
            continue
        normalized.append(section)
        seen.add(section["key"])

    return normalized


def frozen_keys(keys: Iterable[str]) -> frozenset[str]:
    return frozenset(keys)

File: apps/web/src/lib/utils/sectionPreferences.ts
/**
 * Generic configurable-section utilities (#821).
 *
 * Shared merge/normalize/coerce logic behind the Home screen sections (#584)
 * and the Insights page sections (#821). Each caller builds its own set of
 * helpers via {@link createSectionUtils} with its valid keys, default order,
 * and optional locked keys (sections that must always stay enabled).
 */

export interface SectionPreference<K extends string = string> {
  key: K;
  enabled: boolean;
}

export interface SectionUtils<K extends string> {
  DEFAULTS: SectionPreference<K>[];
  merge(stored: SectionPreference<K>[] | null | undefined): SectionPreference<K>[];
  resolveEnabled(sections: SectionPreference<K>[]): SectionPreference<K>[];
  normalizeForSave(sections: SectionPreference<K>[]): SectionPreference<K>[];
  isLocked(key: K): boolean;
}

export function createSectionUtils<K extends string>(config: {
  validKeys: readonly K[];
  defaults: readonly SectionPreference<K>[];
  lockedKeys?: readonly K[];
}): SectionUtils<K> {
  const valid = new Set<string>(config.validKeys);
  const locked = new Set<string>(config.lockedKeys ?? []);
  const defaults = config.defaults;

  function isKey(value: string): value is K {
    return valid.has(value);
  }

  function coerce(raw: unknown): SectionPreference<K> | null {
    if (!raw || typeof raw !== 'object') return null;
    const record = raw as Record<string, unknown>;
    const key = record.key;
    const enabled = record.enabled;
    if (typeof key !== 'string' || !isKey(key.trim())) return null;
    if (typeof enabled !== 'boolean') return null;
    const trimmed = key.trim() as K;
    return { key: trimmed, enabled: locked.has(trimmed) ? true : enabled };
  }

  function merge(stored: SectionPreference<K>[] | null | undefined): SectionPreference<K>[] {
    if (!stored?.length) {
      return defaults.map((section) => ({ ...section }));
    }

    const merged: SectionPreference<K>[] = [];
    const seen = new Set<K>();

    for (const raw of stored) {
      const section = coerce(raw);
      if (!section || seen.has(section.key)) continue;
      merged.push(section);
      seen.add(section.key);
    }

    for (const section of defaults) {
      if (!seen.has(section.key)) {
        merged.push({ ...section });
        seen.add(section.key);
      }
    }

    return merged;
  }

  function resolveEnabled(sections: SectionPreference<K>[]): SectionPreference<K>[] {
    return sections.filter((section) => section.enabled);
  }

  function normalizeForSave(sections: SectionPreference<K>[]): SectionPreference<K>[] {
    const normalized: SectionPreference<K>[] = [];
    const seen = new Set<K>();

    for (const raw of sections) {
      const section = coerce(raw);
      if (!section || seen.has(section.key)) continue;
      normalized.push(section);
      seen.add(section.key);
    }

    return normalized;
  }

  return {
    DEFAULTS: defaults.map((section) => ({ ...section })),
    merge,
    resolveEnabled,
    normalizeForSave,
    isLocked: (key: K) => locked.has(key),
  };
}
